### DISCUSSION TITLE
How can i use existing sensors in berry - do i need to rewrite?

### DISCUSSION THREAD

**Discussion by NiklausS** [arendst/Tasmota#20207] [Q&A] [answered]
Hi,
I'm just reading into berry and tasmota. I have installed Tasmota32 LVGL on a M5Stack and this worked after some fiddling around with the display.ini. I also understand the concept of the 'class MPU6886' - the i2c example. As the **lvgl release has no sensors compiled in**, am i supposed to rewrite all the sensors i need in berry or is there a way to use the existing code together with berry ?
My first aim was to use a INA219 and display the values as well as send them out with mqtt. I have also other well known sensors with native drivers in micropython is there a way to reuse?
thanks for you help
 Nik

**Comment by sfromis** [comment]
No, you're not "supposed" to rewrite anything, but to do a custom build including the feature mix you want. And don't worry, that's not a big task.

**Reply by NiklausS** [reply]
Thanks,
The question of accessing the values from eg. the BMP280 or the INA219 is still not clear.  Also how to set callbacks to existing buttons (physical on 37/38/39 ) ?

I've just noted that a bmp280 is recognized and the values are showed, though it says there are no i2c devices compiled in, on the overview. Strange enough, this sensor was not recognized yesterday with the same setup, and today it was. I use a M5Stack with battery and i first need to program the 'on/off' key to deepsleep as i have not found a shutdown function. I'm coming from the micropython / lua corner of the esp2866 / esp32. So building is not an issue. - BUT, i don't want to build if not needed - as the beauty of tasmota are the clean builds and upgradability with OTA, plus a stable base. Once installed, they just work.

**Reply by sfromis** [reply]
Accessing sensor values is like this, to get a `map` reflecting the usual sensor JSON.
```
import json
var values = json.load(tasmota.read_sensors())
```
This is when you are running a build where the relevant sensor drivers are already included. The tool [decode-status](https://github.com/tasmota/decode-status/) can give a list of what features are included in the current build.

And while a precompiled display-oriented build (`tasmota32-lvgl.bin`) does provide *some* sensors, this is by far not a wide selection, and does not include the two ones you asked about. Hence, you do need the standard and pretty "clean" procedure of doing a custom build for the feature mix you want.

About your "strange" things, not clear exactly what binaries you were using and what you saw.

**Reply by NiklausS** [reply]
Thanks again for the detailed answer.

i found in the meantime this (minimal) solution to get a callback with a value:
```
def rule_temp(value)
    print("The temp is bright", value)
end
tasmota.add_rule("BMP280#Temperature<27.3", rule_temp )
```
allowing me straight forward to display the value on the display by adding some lv lines. 

the **tasmota.read_sensors()** did'nt caught me attention on first reading ;-) - seems to be what i was looking for.

the 'strange' thing is: after i had the display running i attached a BMP280 sensor to the M5Stack - quit simple. And i expected to see the temp values on the tasmota main page. But there was nothing, even after several restarts - although the i2cscan showed the it correct. So i check the build overview and found that no i2c sensors are included in the display builds. Today then - the sensor showed up as expected - and all in between was a DeepSleepTime 860000 to not deplete the battery and a reset today.

**Reply by sfromis** [reply, marked as answer]
Again **some** I2C sensors *are* included in normal LVGL builds (which are preferred for LVGL), but only a small selection. Still does include the BME280, which is part of driver number 10, if enabled `{"I2CDriver":"7,9,10,58,59,77"}`,

If you instead use `tasmota32-display.bin` (not meant for LVGL) this also includes a different small selection if I2C drivers, not including BME280. `{"I2CDriver":"3,4,5,6,38,39,47"}`

Hence it matters a lot what *specific* build you were using, when you saw or not saw something. **And**, the intention is that you do a custom build for the feature mix you want.
